### PR TITLE
Script to setup integration tests

### PR DESCRIPTION
**Pull request by piotrkwiecinski**
Script to setup env for integration tests

It creates a integration database and generates `install-config-mysql.php` based on the .dist version.

### CODE REVIEW

File: compose/bin/setup-integration-tests [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+source env/db.env
+
+REAL_SRC=$(cd -P "src" && pwd)
+MYSQL_CONFIG=dev/tests/integration/etc/install-config-mysql.php
+INTEGRATION_DB=magento_integration_tests
+CREATE_DB="bin/clinotty mysqladmin -hdb -uroot -p${MYSQL_ROOT_PASSWORD} create ${INTEGRATION_DB}"  && echo "Database ${INTEGRATION_DB} created."
```
Comment (piotrkwiecinski) [inline compose/bin/setup-integration-tests:7]: I couldn't find cleaner way to handle exit code from command run on a container. Suggestions are welcome.

File: compose/bin/setup-integration-tests [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+bin/clinotty mysql -hdb -uroot -p${MYSQL_ROOT_PASSWORD} ${INTEGRATION_DB} -e exit &> /dev/null || $CREATE_DB
+
+sed -e "s/'db-host' => 'localhost'/'db-host' => 'db'/" \
```
Comment (piotrkwiecinski) [inline compose/bin/setup-integration-tests:10]: Are we ok with using root user or do we want to create a separate user?
We could put credentials in `env/db.env` if you prefer.

File: compose/bin/setup-integration-tests [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  -e "s/'db-password' => '123123q'/'db-password' => '${MYSQL_ROOT_PASSWORD}'/" \
+  -e "s/'amqp-host' => 'localhost'/'amqp-host' => 'rabbitmq'/" \
+  src/dev/tests/integration/etc/install-config-mysql.php.dist > src/dev/tests/integration/etc/install-config-mysql.php
+
+# Workaround until coping nested files works as expected
+docker cp $REAL_SRC/$MYSQL_CONFIG $(docker-compose ps -q phpfpm|awk '{print $1}'):/var/www/html/$MYSQL_CONFIG
```
[the remark this answers is not in the document]
Comment (markshust) [reply compose/bin/setup-integration-tests:16]: #295 has been merged

File: compose/bin/setup-integration-tests (hunk 1 above)
Comment (markshust) [reply compose/bin/setup-integration-tests:7]: would this line be better?

`CREATE_DB="bin/clinotty mysqladmin -hdb -uroot -p${MYSQL_ROOT_PASSWORD} create ${INTEGRATION_DB}" || exit`

File: compose/bin/setup-integration-tests (hunk 2 above)
Comment (markshust) [reply compose/bin/setup-integration-tests:10]: I think `root` user is fine, as most of these scripts are meant for development. However, I think putting creds in `env/db.env` are the right way to go:

```
...
MYSQL_INTEGRATION_ROOT_PASSWORD=magento
MYSQL_INTEGRATION_DATABASE=magento_integration
MYSQL_INTEGRATION_USER=magento
MYSQL_INTEGRATION_PASSWORD=magento
etc
```

Note renaming `INTEGRATION_DB ` to `MYSQL_INTEGRATION_DATABASE` and moving that into the env.db file. I'd also keep all the users/creds/etc `magento` for simplicity sake.

**Comment by piotrkwiecinski** [comment]
@markshust it should be ready for another round. Credentials are used from env/db.env. 
I hope I've cleaned up the script.

**Review by markshust — APPROVED** [review]

**Comment by markshust** [comment]
@piotrkwiecinski I merged in because there aren't any breaking updates here, but I got this error when running it:

```
m240$ bin/magento dev:test:run integration

---- /var/www/html/dev/tests/integration> /usr/local/bin/php /var/www/html/./vendor/phpunit/phpunit/phpunit  

Exception: Could not connect to the Amqp Server.
SQLSTATE[HY000] [2002] No such file or directory

In InstallCommand.php line 274:

  [InvalidArgumentException]
  Parameter validation failed


Exception trace:
  at /var/www/html/setup/src/Magento/Setup/Console/Command/InstallCommand.php:274
 Magento\Setup\Console\Command\InstallCommand->initialize() at /var/www/html/vendor/symfony/console/Command/Command.php:221
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:1000
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:271
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/magento/framework/Console/Cli.php:115
 Magento\Framework\Console\Cli->doRun() at /var/www/html/vendor/symfony/console/Application.php:147
 Symfony\Component\Console\Application->run() at /var/www/html/bin/magento:23

setup:install [--amqp-host AMQP-HOST] [--amqp-port AMQP-PORT] [--amqp-user AMQP-USER] [--amqp-password AMQP-PASSWORD] [--amqp-virtualhost AMQP-VIRTUALHOST] [--amqp-ssl AMQP-SSL] [--amqp-ssl-options AMQP-SSL-OPTIONS] [--enable-debug-logging ENABLE-DEBUG-LOGGING] [--enable-syslog-logging ENABLE-SYSLOG-LOGGING] [--backend-frontname BACKEND-FRONTNAME] [--consumers-wait-for-messages CONSUMERS-WAIT-FOR-MESSAGES] [--key KEY] [--db-host DB-HOST] [--db-na...
```

**Comment by markshust** [comment]
Later in the error I also see this...

```
#2 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(56): include_once('/var/www/html/d...')
#3 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(44): PHPUnit\Util\FileLoader::load('/var/www/html/d...')
#4 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(552): PHPUnit\Util\FileLoader::checkAndLoad('/var/www/html/d...')
#5 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(360): PHPUnit\TextUI\Command->handleBootstrap('/var/www/html/d...')
#6 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(76): PHPUnit\TextUI\Command->handleArguments(Array)
#7 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(68): PHPUnit\TextUI\Command->run(Array, true)
#8 /var/www/html/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#9 {main}

Fatal error: Uncaught Exception: sh: 1: tasklist.exe: not found in /var/www/html/vendor/magento/framework/Shell.php:63
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(92): Magento\Framework\Shell->execute('tasklist.exe /f...', Array)
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(58): Magento\TestFramework\Helper\Memory->_getWinProcessMemoryUsage(20)
#2 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(127): Magento\TestFramework\Helper\Memory->getRealMemoryUsage()
#3 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(59): Magento\TestFramework\MemoryLimit->_getUsage()
#4 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Bootstrap/Memory.php(50): Magento\TestFramework\MemoryLimit->printStats()
#5 [internal function]: Magento\TestFramework\Bootstrap\Memory->displayStats()
#6 {main}

Next Magento\Framework\Exception\LocalizedException: Command re in /var/www/html/vendor/magento/framework/Shell.php on line 64
```

**Comment by piotrkwiecinski** [comment]
@markshust tasklist exec should be resolved by: https://github.com/markshust/docker-magento/commit/1b71f7f956cd7bd3ce69ba8d47604171842878a1 but I'm not sure if you update an image after that change was merged.

I'll have a look a the amqp tomorrow or on saturday. That branch is a bit old :).

**Comment by markshust** [comment]
Ah thanks -- I'll test with the new image build

**Comment by markshust** [comment]
Just for reference, here's output with the new PHP image:

```
m240$ bin/magento dev:test:run integration

---- /var/www/html/dev/tests/integration> /usr/local/bin/php /var/www/html/./vendor/phpunit/phpunit/phpunit  

Exception: Could not connect to the Amqp Server.
SQLSTATE[HY000] [2002] No such file or directory

In InstallCommand.php line 274:

  [InvalidArgumentException]
  Parameter validation failed


Exception trace:
  at /var/www/html/setup/src/Magento/Setup/Console/Command/InstallCommand.php:274
 Magento\Setup\Console\Command\InstallCommand->initialize() at /var/www/html/vendor/symfony/console/Command/Command.php:221
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:1000
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:271
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/magento/framework/Console/Cli.php:115
 Magento\Framework\Console\Cli->doRun() at /var/www/html/vendor/symfony/console/Application.php:147
 Symfony\Component\Console\Application->run() at /var/www/html/bin/magento:23

setup:install [--amqp-host AMQP-HOST] [--amqp-port AMQP-PORT] [--amqp-user AMQP-USER] [--amqp-password AMQP-PASSWORD] [--amqp-virtualhost AMQP-VIRTUALHOST] [--amqp-ssl AMQP-SSL] [--amqp-ssl-options AMQP-SSL-OPTIONS] [--enable-debug-logging ENABLE-DEBUG-LOGGING] [--enable-syslog-logging ENABLE-SYSLOG-LOGGING] [--backend-frontname BACKEND-FRONTNAME] [--consumers-wait-for-messages CONSUMERS-WAIT-FOR-MESSAGES] [--key KEY] [--db-host DB-HOST] [--db-name DB-NAME] [--db-user DB-USER] [--db-engine DB-ENGINE] [--db-password DB-PASSWORD] [--db-prefix DB-PREFIX] [--db-model DB-MODEL] [--db-init-statements DB-INIT-STATEMENTS] [-s|--skip-db-validation] [--http-cache-hosts HTTP-CACHE-HOSTS] [--db-ssl-key DB-SSL-KEY] [--db-ssl-cert DB-SSL-CERT] [--db-ssl-ca DB-SSL-CA] [--db-ssl-verify] [--session-save SESSION-SAVE] [--session-save-redis-host SESSION-SAVE-REDIS-HOST] [--session-save-redis-port SESSION-SAVE-REDIS-PORT] [--session-save-redis-password SESSION-SAVE-REDIS-PASSWORD] [--session-save-redis-timeout SESSION-SAVE-REDIS-TIMEOUT] [--session-save-redis-persistent-id SESSION-SAVE-REDIS-PERSISTENT-ID] [--session-save-redis-db SESSION-SAVE-REDIS-DB] [--session-save-redis-compression-threshold SESSION-SAVE-REDIS-COMPRESSION-THRESHOLD] [--session-save-redis-compression-lib SESSION-SAVE-REDIS-COMPRESSION-LIB] [--session-save-redis-log-level SESSION-SAVE-REDIS-LOG-LEVEL] [--session-save-redis-max-concurrency SESSION-SAVE-REDIS-MAX-CONCURRENCY] [--session-save-redis-break-after-frontend SESSION-SAVE-REDIS-BREAK-AFTER-FRONTEND] [--session-save-redis-break-after-adminhtml SESSION-SAVE-REDIS-BREAK-AFTER-ADMINHTML] [--session-save-redis-first-lifetime SESSION-SAVE-REDIS-FIRST-LIFETIME] [--session-save-redis-bot-first-lifetime SESSION-SAVE-REDIS-BOT-FIRST-LIFETIME] [--session-save-redis-bot-lifetime SESSION-SAVE-REDIS-BOT-LIFETIME] [--session-save-redis-disable-locking SESSION-SAVE-REDIS-DISABLE-LOCKING] [--session-save-redis-min-lifetime SESSION-SAVE-REDIS-MIN-LIFETIME] [--session-save-redis-max-lifetime SESSION-SAVE-REDIS-MAX-LIFETIME] [--session-save-redis-sentinel-master SESSION-SAVE-REDIS-SENTINEL-MASTER] [--session-save-redis-sentinel-servers SESSION-SAVE-REDIS-SENTINEL-SERVERS] [--session-save-redis-sentinel-verify-master SESSION-SAVE-REDIS-SENTINEL-VERIFY-MASTER] [--session-save-redis-sentinel-connect-retires SESSION-SAVE-REDIS-SENTINEL-CONNECT-RETIRES] [--cache-backend CACHE-BACKEND] [--cache-backend-redis-server CACHE-BACKEND-REDIS-SERVER] [--cache-backend-redis-db CACHE-BACKEND-REDIS-DB] [--cache-backend-redis-port CACHE-BACKEND-REDIS-PORT] [--cache-backend-redis-password CACHE-BACKEND-REDIS-PASSWORD] [--cache-backend-redis-compress-data CACHE-BACKEND-REDIS-COMPRESS-DATA] [--cache-backend-redis-compression-lib CACHE-BACKEND-REDIS-COMPRESSION-LIB] [--cache-id-prefix CACHE-ID-PREFIX] [--allow-parallel-generation] [--page-cache PAGE-CACHE] [--page-cache-redis-server PAGE-CACHE-REDIS-SERVER] [--page-cache-redis-db PAGE-CACHE-REDIS-DB] [--page-cache-redis-port PAGE-CACHE-REDIS-PORT] [--page-cache-redis-password PAGE-CACHE-REDIS-PASSWORD] [--page-cache-redis-compress-data PAGE-CACHE-REDIS-COMPRESS-DATA] [--page-cache-redis-compression-lib PAGE-CACHE-REDIS-COMPRESSION-LIB] [--page-cache-id-prefix PAGE-CACHE-ID-PREFIX] [--lock-provider LOCK-PROVIDER] [--lock-db-prefix LOCK-DB-PREFIX] [--lock-zookeeper-host LOCK-ZOOKEEPER-HOST] [--lock-zookeeper-path LOCK-ZOOKEEPER-PATH] [--lock-file-path LOCK-FILE-PATH] [--base-url BASE-URL] [--language LANGUAGE] [--timezone TIMEZONE] [--currency CURRENCY] [--use-rewrites USE-REWRITES] [--use-secure USE-SECURE] [--base-url-secure BASE-URL-SECURE] [--use-secure-admin USE-SECURE-ADMIN] [--admin-use-security-key ADMIN-USE-SECURITY-KEY] [--admin-user [ADMIN-USER]] [--admin-password [ADMIN-PASSWORD]] [--admin-email [ADMIN-EMAIL]] [--admin-firstname [ADMIN-FIRSTNAME]] [--admin-lastname [ADMIN-LASTNAME]] [--search-engine SEARCH-ENGINE] [--elasticsearch-host ELASTICSEARCH-HOST] [--elasticsearch-port ELASTICSEARCH-PORT] [--elasticsearch-enable-auth ELASTICSEARCH-ENABLE-AUTH] [--elasticsearch-username ELASTICSEARCH-USERNAME] [--elasticsearch-password ELASTICSEARCH-PASSWORD] [--elasticsearch-index-prefix ELASTICSEARCH-INDEX-PREFIX] [--elasticsearch-timeout ELASTICSEARCH-TIMEOUT] [--cleanup-database] [--sales-order-increment-prefix SALES-ORDER-INCREMENT-PREFIX] [--use-sample-data] [--enable-modules [ENABLE-MODULES]] [--disable-modules [DISABLE-MODULES]] [--convert-old-scripts [CONVERT-OLD-SCRIPTS]] [-i|--interactive] [--safe-mode [SAFE-MODE]] [--data-restore [DATA-RESTORE]] [--dry-run [DRY-RUN]] [--magento-init-params MAGENTO-INIT-PARAMS]
 in /var/www/html/vendor/magento/framework/Shell.php:63
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Application.php(515): Magento\Framework\Shell->execute('/usr/local/bin/...', Array)
#1 /var/www/html/dev/tests/integration/framework/bootstrap.php(89): Magento\TestFramework\Application->install(true)
#2 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(56): include_once('/var/www/html/d...')
#3 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(44): PHPUnit\Util\FileLoader::load('/var/www/html/d...')
#4 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(552): PHPUnit\Util\FileLoader::checkAndLoad('/var/www/html/d...')
#5 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(360): PHPUnit\TextUI\Command->handleBootstrap('/var/www/html/d...')
#6 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(76): PHPUnit\TextUI\Command->handleArguments(Array)
#7 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(68): PHPUnit\TextUI\Command->run(Array, true)
#8 /var/www/html/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#9 {main}

Next Magento\Framework\Exception\LocalizedException: Command returned non-zero exit code:
`/usr/local/bin/php -f '/var/www/html/bin/magento' setup:install -vvv --db-host='localhost' --db-user='root' --db-password='123123q' --db-name='magento_integration_tests' --backend-frontname='backend' --admin-user='user' --admin-password='password1' --admin-email='admin@example.com' --admin-firstname='firstname' --admin-lastname='lastname' --amqp-host='localhost' --amqp-port='5672' --amqp-user='guest' --amqp-password='guest' --magento-init-params='MAGE_DIRS[etc][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/etc&MAGE_DIRS[var][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var&MAGE_DIRS[var_export][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/export&MAGE_DIRS[media][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/pub/media&MAGE_DIRS[static][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/pub/static&MAGE_DIRS[view_preprocessed][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/view_preprocessed/pub/static&MAGE_DIRS[code][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/generated/code&MAGE_DIRS[cache][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/cache&MAGE_DIRS[log][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/log&MAGE_DIRS[session][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/session&MAGE_DIRS[tmp][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/tmp&MAGE_DIRS[upload][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/var/upload&MAGE_DIRS[pub][path]=/var/www/html/dev/tests/integration/tmp/sandbox-0-9ceb766674d7a6230ff26621b8197274e42637fd91c0e84ddf9d67b3cab3477a/pub&MAGE_MODE=developer' 2>&1` in /var/www/html/vendor/magento/framework/Shell.php:64
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Application.php(515): Magento\Framework\Shell->execute('/usr/local/bin/...', Array)
#1 /var/www/html/dev/tests/integration/framework/bootstrap.php(89): Magento\TestFramework\Application->install(true)
#2 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(56): include_once('/var/www/html/d...')
#3 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(44): PHPUnit\Util\FileLoader::load('/var/www/html/d...')
#4 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(552): PHPUnit\Util\FileLoader::checkAndLoad('/var/www/html/d...')
#5 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(360): PHPUnit\TextUI\Command->handleBootstrap('/var/www/html/d...')
#6 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(76): PHPUnit\TextUI\Command->handleArguments(Array)
#7 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(68): PHPUnit\TextUI\Command->run(Array, true)
#8 /var/www/html/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#9 {main}

=== Memory Usage System Stats ===
Memory usage (OS):	37.44M (374.38% of 10.00M reported by PHP)
Estimated memory leak:	27.44M (73.29% of used memory)

----------------------------------------------------------------------
FAILED - 1 of 1:
 - /var/www/html/dev/tests/integration> /usr/local/bin/php /var/www/html/./vendor/phpunit/phpunit/phpunit 
```

**Pull request merged**